### PR TITLE
[MPS] Add MPS backend plugin for GPU computing on Macbook

### PR DESCRIPTION
**Pull request by lishicheng1996**
This pull request adds the MPS backend plugin to PaddlePaddle, allowing it to use GPU computing on Macbook. The first kernel, Softmax, has been added to the plugin. 
1. The plugin can be compiled, installed, and used by following the instructions in the README.md  
2. PaddlePaddle needs to expose capi on macOS. A pull request will be submitted soon. 
3. Note that the dylib in this plugin currently requires developer signing.
4. Currently all workloads are commited to the default stream, as same as pytorch does.

### CODE REVIEW

File: backends/mps/runtime/runtime.cc [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+C_Status StreamWaitEvent(const C_Device device,
+                         C_Stream stream,
+                         C_Event event) {
+  return C_SUCCESS;
+}
+
+C_Status VisibleDevices(size_t* devices) { return C_SUCCESS; }
+
+C_Status DeviceMemStats(const C_Device device,
+                        size_t* total_memory,
```
Comment (ronny1996) [inline backends/mps/runtime/runtime.cc:150]: mps没有设备内存统计的接口吗？

**Review by ronny1996 — APPROVED** [review]
LGTM

File: backends/mps/runtime/runtime.cc (hunk 1 above)
Comment (lishicheng1996) [reply backends/mps/runtime/runtime.cc:150]: 内存统计这块我还没弄，目前也还不知道有没有原生api可以直接统计MTLBuffer的内存占用。pytorch是维护了MTLBuffer的内存池可以统计。硬件角度M1是unified memory，cpu和gpu都是放一块内存上，数据结构的封装不同。

**Comment by engineer1109** [comment]
不太清楚MacOS.
MTLBuffer能否当指针一样使用？
paddle上是需要支持指针访问，包括char* ptr = p + 100; 这样的操作。
Paddle上的内存分配策略是有内存复用的，Alloc<T>申请的内存，可能是之前申请内存中的一小部分。

我这里对于OpenCL，可以明确CLBuffer是alloc_memory 不能返回强转的void*, 必须实现一个指针访问的内存池。
你在跑PaddleInference的时候就会发现这个问题。

**Comment by lishicheng1996** [comment]
> MTLBuffer是不能直接当指针使用的。是否可以先关闭paddle的内存复用呀？我可以看看OpenCL是怎么处理的吗？需要的话我也实现一个

**Comment by lishicheng1996** [comment]
> 不太清楚MacOS. MTLBuffer能否当指针一样使用？ paddle上是需要支持指针访问，包括char* ptr = p + 100; 这样的操作。 Paddle上的内存分配策略是有内存复用的，Alloc申请的内存，可能是之前申请内存中的一小部分。
> 
> 我这里对于OpenCL，可以明确CLBuffer是alloc_memory 不能返回强转的void*, 必须实现一个指针访问的内存池。 你在跑PaddleInference的时候就会发现这个问题。

**Comment by lishicheng1996** [comment]
> 

谢谢提醒，MTLBuffer是不能直接当作指针使用的。但是paddle在使用时可以选择关掉内存复用的，可以参考[这里](https://github.com/PaddlePaddle/PaddleCustomDevice/issues/244)。后续的开发里有需要的话，我这边再给MTLBuffer维护内存池 @engineer1109

**Comment by engineer1109** [comment]
@lishicheng1996 关闭内存复用我到现在也没法解决，原因是里面有很多设计都是以cuda和cpu考虑的，既然要求返回void*指针，那么paddle上层肯定是当指针来用的。

OpenCL是我们公司闭源实现，目前没考虑开放。

目前，OpenCL实现Paddle的思路有两个方案
1. 使用clSVMAlloc分配虚拟内存的指针，可以直接当cuda指针用。
2. 自己建立虚拟内存池，建立多个超大的Buffer，每个1G大小，记录每个DenseTensor对内存使用的情况。
把指针和Buffer的origin (buf开头偏移量)和size （Tensor占用的大小） 建立转换关系。

例如void* ptr = (size_t ID, size_t origin)
ID是内存池的序号，只有一块的话就是0, origin是内存池的起始位置， 这样可以模拟出指针的效果。

这个问题我之前在PFCC的PPT分享讲过，分析过OpenCL Vulkan OpenGL实现的难度。

我这里的提醒是让你少走弯路，你看intel为什么用device_alloc(相当于clSVMAlloc), 这个问题只有做过的人知道，百度对非指针访问的图形API是不了解的。也不可能为此设计。

**Comment by engineer1109** [comment]
@lishicheng1996 无论有没有内存优化图，总有奇怪的地方会涉及指针偏移，当你完整实现一个yolov7就知道了。我们这里大部分网络都已经完整支持了。所以才能了解这些情况。

**Comment by lishicheng1996** [comment]
> @lishicheng1996 无论有没有内存优化图，总有奇怪的地方会涉及指针偏移，当你完整实现一个yolov7就知道了。我们这里大部分网络都已经完整支持了。所以才能了解这些情况。

@engineer1109 嗯嗯，真诚的感谢您的提醒和宝贵的经验。这个pr的设计是不完备的，我后续尝试支持stable diffusion模型。过程中肯定是需要做出调整和优化。
MTLBuffer是支持偏移量的，可以参考[这里](https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443126-setbuffer?language=objc#:~:text=the%20argument%20table.-,offset,-Where%20the%20data)，但是不像cuda那么方便，可以像裸指针一样直接加减偏移量。
